### ISSUE TITLE
Click Trigger Uses the Datapoints of a Drawing Instead of Its Bounding Box

### ISSUE THREAD

**Issue by AnthonyWinder** [League-of-Foundry-Developers/fvtt-module-trigger-happy#86]
Hello all. I'm afraid that I'm new enough to Javascipt and Github and I don't know the appropriate method for submitting an improvement to the library, but I've put together some code that allows the user to use a drawing of any shape as a click trigger (using the geometry of the shape instead of its bounding box). I use it in my own project to make interactive maps since my irregularly-shaped continents have overlapping bounding boxes but are easy to outline with the freehand and polygon tools. All that is required in order to implement the improvement is to replace the _placeableContains method in trigger.js with the following code:

```
  _placeableContains(placeable, position) {
    // The placeable could have many types: TokenDocument, DrawingDocument, NoteDocument, JournalEntry, and perhaps others
    // If I understand correctly, however, only TokenDocuments and DrawingDocuments have an 'area' to consider for collisions

    if (placeable instanceof TokenDocument) {
      // TokenDocument.data reports width and height in terms of grid units. We could calculate the size in pixels manually, but
      // TokenDocument.object already reports the size and position of the token in pixels, so we will use it instead.
      //
      // TokenDocument.object.w and .h report the size of the token object according to its boundaries on the grid (orange rectangle on hover) 
      // while .width and .height refer to the token image, which takes into account the image scale as well. For the purpose of clicks and token
      // collisions, we will consider scaling and rotation of the token image to be purely aesthetic, and work off of the token's grid dimensions
      let x = placeable.object.x
	  let y = placeable.object.y
	  let width = placeable.object.w
	  let height = placeable.object.h

	  // For both Tokens and Drawings, we will start by determining if 'position' is inside the placeable's bounding box
      // Since tokens have rectangular boundaries, we don't need to perform any other calculations
      return (Number.between(position.x, x, x + width) && Number.between(position.y, y, y + height))
    }

    else if (placeable instanceof DrawingDocument) {
      // For a DrawingDocument, the width and height of the contents are stored in DrawingDocument.data
      // the .data member also includes .points, which contains the coordinates of the drawing's vertices relative to the drawing's origin
      let x = placeable.data.x
	  let y = placeable.data.y
 	  let width = placeable.data.width
	  let height = placeable.data.height
	  // Possible drawing types: (r)ectangle, circl(e), (p)olygon, (f)reehand
	  let type = placeable.data.type

	  if (placeable.data.rotation != 0) {
	    // It looks like Foundry applies the rotation to the image drawn on the canvas, but not to the position or size of the DrawingData
        // Instead of rotating the entire DrawingData to match what is rendered on the canvas, we can just inverse-rotate position
        let drawing_center = [x + 0.5*width, y + 0.5*height]
        position = {
          x: Math.cos(-placeable.data.rotation*Math.PI/180) * (position.x - drawing_center[0]) - Math.sin(-placeable.data.rotation*Math.PI/180) * (position.y - drawing_center[1]) + drawing_center[0],
          y: Math.sin(-placeable.data.rotation*Math.PI/180) * (position.x - drawing_center[0]) + Math.cos(-placeable.data.rotation*Math.PI/180) * (position.y - drawing_center[1]) + drawing_center[1]
        }
      }

	  // For both Tokens and Drawings, we will start by determining if 'position' is inside the placeable's bounding box
      if (Number.between(position.x, x, x + width) && Number.between(position.y, y, y + height)) {
        // If the position is within the bounding box, then we can perform additional tests to see if it is within the drawing's geometry
        if (type == "r") {
          // A rectangle is its own bounding box, so we've already determined that it contains the position
          return true
        } 
        else if (type == "e") {
          // All points inside an ellipse satisfy the following inequality
          return ((position.x - x - 0.5*width)**2 * (0.5*height)**2 + (position.y - y - 0.5*height)**2 * (0.5*width)**2) <= ((0.5*width)**2 * (0.5*height)**2)
        }
        else if (type == "p" || type == "f") {
          // Point and freehand drawings have point data that we can use to perform a point inclusion in polygon test as described in https://wrf.ecse.rpi.edu/Research/Short_Notes/pnpoly.html
          // To get the pixel coordinates of the vertices, we have to add the drawing origin
	      let vertices = placeable.data.points.map(point => [point[0]+x, point[1]+y])
          let isInside = false
          let i = 0, j = vertices.length - 1;
          for (i, j; i < vertices.length; j = i++) {
              if ( (vertices[i][1] > position.y) != (vertices[j][1] > position.y) && (position.x < (vertices[j][0] - vertices[i][0]) * (position.y - vertices[i][1]) / (vertices[j][1] - vertices[i][1]) + vertices[i][0])) {
                isInside = !isInside;
              }
          }
          return isInside
        }
        else {
          // If this runs, then foundry added a new drawing type. Guess we'll default to just using the bounding box
          return true
        }
      }
      else {
        // If the position is outside the bounding box, then we know immediately that the drawing does not contain it
        return false
      }
    }
    
    else {
      // Other types of placeables don't have an area that could contain the position
      return false
    }
  }
```

The elliptical and point inclusion tests only run for points inside the drawing's minimum bounding box which means that there's only a very slight increase in compute time. I think I've also addressed the major concerns related to token scaling. That said, this is my first attempt at contributing code to a Foundry module and I would definitely welcome constructive criticism. On that point, I'm pretty certain that filing a bug report is not the proper way to contribute to a Github project, so if anyone could fill me in on how I *should* have submitted this code, I would greatly appreciate it.

**Comment by p4535992** [comment]
ty for the contribution, sadly i'm a little busy lately, i will check out your solution this week end.

**Comment by AnthonyWinder** [comment]
> ty for the contribution, sadly i'm a little busy lately, i will check out your solution this week end.

You're very welcome. It's a busy time for us all; there is no rush. I can already use my changes locally so I have no personal stake in getting the code reviewed quickly. I was just thinking that others may find this feature useful in the (indeterminably distant) future.

**Comment by p4535992** [comment]
hey @AnthonyWinder here the beta:
`https://github.com/p4535992/fvtt-module-trigger-happy/releases/download/v0.8.33/module.json`
let me know if it work and i will prepare a new release.

**Comment by AnthonyWinder** [comment]
> hey @AnthonyWinder here the beta: `https://github.com/p4535992/fvtt-module-trigger-happy/releases/download/v0.8.33/module.json` let me know if it work and i will prepare a new release.

That was quick!

Placing the pre-existing code in the 'else' clause that deals with other document types was a smart move. Only issue I found was some indenting on lines 1218-1222 and 1261, but that looks like it was my fault - extra tabs got added when I pasted the code above. The code runs perfectly for my use case, which is creating clickable drawings. I wasn't able to test drawings as collision triggers for moving tokens, but I would assume that if one works, the other must too.
